exercises/word-search/example.js

```diff
@@ -0,0 +1,146 @@
+class WordSearch {
+  constructor(grid) {
+    this.grid = grid;
+  }
+
+  find(words) {
+    return words
+      .map(word => ({[word]: findWordInAnyDirection(word, this.grid)}))
+      .reduce((acc, oneWord) => Object.assign(acc, oneWord), {});
+  }
+}
+
+function findWordInAnyDirection(word, grid) {
+  return searchHorizontally({word, grid})
+    || flipCoordinates(searchHorizontally({word, grid: flipGrid(grid)}))
+    || searchDiagonally({word, grid})
+    || searchDiagonally({word, grid, fromTop: false});
+}
+
+function searchHorizontally({word, grid}) {
+  let rowIndex = 0;
+  let startCol;
+  let start;
+  let end;
+
+  const getCoords = () => [[rowIndex + 1, startCol], [rowIndex + 1, startCol + word.length - 1]];
+
+  const getStartCol = (word) => 1 + grid[rowIndex].indexOf(word);
+
+  while (rowIndex < grid.length) {
+    startCol = getStartCol(word);
+    if (startCol) {
+      [start, end] = getCoords();
+    } else {
+      startCol = getStartCol([...word].reverse().join(''));
+      if (startCol) {
+        [end, start] = getCoords();
+      }
+    }
+    if (start && end) {
+      return {start, end};
+    }
+    rowIndex++;
+  }
+}
+
+function flipCoordinates(coords) {
+  if (!coords) {
+    return undefined;
+  }
+  return {
+    start: coords.start.reverse(),
+    end:   coords.end.reverse()
+  };
+}
+
+function flipGrid(grid) {
+  return [...grid[0]]
+    .map((col, c) => grid.map((row, r) => grid[r][c]))
+    .map(row => row.join(''))
+}
+
+function searchDiagonally({word, grid, isReversed = false, fromTop = true}) {
```

Comment:
I'm not a big fan of optional arguments. But I don't have a better suggestion.

To search horizontally in both directions, you implemented a `flipGrid` function. Would it be useful to search diagonally? Would there be a similar function to search diagonally by *flipping* the grid? I'm just thinking outloud, I'm not asking for implementing it.

Reply:
@rchavarria I had thought about rotating the grid 45 degrees to do the diagonal search, but I couldn't quite figure out how to make that happen.

If it were possible to do that, then you would only need to search in one direction: just rotate the grid each time, and correctly transform the coordinates back to the original system.

Maybe some enterprising exercism user will figure that one out...?

Reply:
Oh, I see. 45 degrees rotation? Interesting point, but not sure how to reflect that. And, I don't know if the exercise forces the grid to be square. Let's keep it as it is.